### PR TITLE
Prevent ArgumentOutOfRangeException during Inline Rename navigation

### LINKED ISSUE #7660: VS2015.1: ArgumentOutOfRangeException while inline renaming in razor
With `Index.cshtml` looking like

``` cshtml
@using ArgumentOutOfRangeTest.Controllers
@model Some
@Html.GetTest().TestMethod(x =>
{
    x.SomeMethod(new { Controller = "Test", Action = nameof(TestController.Other) });
}).GetHtml()
```

I set the cursor to `Other`, pressed `F2` and got an error MessageBox telling me about a ArgumentOutOfRangeException for parametername position.
Pressing `F2` again, it did the same error again.
Attaching Visual Studio got me this stacktrace:

```
ArgumentOutOfRangeException: param: position
   at Microsoft.VisualStudio.Text.Implementation.BinaryStringRebuilder.GetLineNumberFromPosition(Int32 position)
   at Microsoft.VisualStudio.Text.Implementation.TextSnapshot.GetLineNumberFromPosition(Int32 position)
   at Microsoft.CodeAnalysis.Text.Extensions.SnapshotSourceText.LineInfo.IndexOf(Int32 position)
   at Microsoft.CodeAnalysis.Text.Extensions.SnapshotSourceText.LineInfo.GetLineFromPosition(Int32 position)
   at Microsoft.CodeAnalysis.Shared.Extensions.SourceTextExtensions.GetLineAndOffset(SourceText text, Int32 position, Int32& lineNumber, Int32& offset)
   at Microsoft.VisualStudio.LanguageServices.Implementation.Extensions.SourceTextExtensions.GetVsTextSpanForSpan(SourceText text, TextSpan textSpan)
   at Microsoft.VisualStudio.LanguageServices.Implementation.VisualStudioDocumentNavigationService.CanNavigateToSpan(Workspace workspace, DocumentId documentId, TextSpan textSpan)
   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.AbstractEditorInlineRenameService.GetRenameInfo(IEnumerable`1 refactorNotifyServices, Document document, SyntaxToken triggerToken, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.AbstractEditorInlineRenameService.GetRenameInfo(Document document, Int32 position, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.AbstractEditorInlineRenameService.GetRenameInfoAsync(Document document, Int32 position, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.InlineRenameService.StartInlineSession(Document document, TextSpan textSpan, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.RenameCommandHandler.ExecuteRenameWorker(RenameCommandArgs args, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.RenameCommandHandler.<>c__DisplayClass37_0.<ExecuteCommand>b__0(IWaitContext waitContext)
   at Microsoft.VisualStudio.LanguageServices.Implementation.Utilities.VisualStudioWaitIndicator.Wait(String title, String message, Boolean allowCancel, Action`1 action)
   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.RenameCommandHandler.ExecuteCommand(RenameCommandArgs args, Action nextHandler)
   at Microsoft.CodeAnalysis.Editor.Implementation.Commands.CommandHandlerService.ExecuteHandlers[T](IList`1 commandHandlers, T args, Action lastHandler)
   at Microsoft.CodeAnalysis.Editor.Implementation.Commands.CommandHandlerService.Microsoft.CodeAnalysis.Editor.ICommandHandlerService.Execute[T](IContentType contentType, T args, Action lastHandler)
   at Microsoft.VisualStudio.LanguageServices.Implementation.AbstractOleCommandTarget.ExecuteRename(ITextBuffer subjectBuffer, IContentType contentType, Action executeNextCommandTarget)
   at Microsoft.VisualStudio.LanguageServices.Implementation.AbstractOleCommandTarget.ExecuteVisualStudio2000(Guid& pguidCmdGroup, UInt32 commandId, UInt32 executeInformation, IntPtr pvaIn, IntPtr pvaOut, ITextBuffer subjectBuffer, IContentType contentType)
   at Microsoft.VisualStudio.LanguageServices.Implementation.AbstractOleCommandTarget.Exec(Guid& pguidCmdGroup, UInt32 commandId, UInt32 executeInformation, IntPtr pvaIn, IntPtr pvaOut)
   at Microsoft.VisualStudio.Html.Interop.OleToCommandTargetShim.Invoke(Guid group, Int32 id, Object inputArg, Object& outputArg)
   at Microsoft.Html.Editor.Commands.HtmlMainController.Invoke(Guid group, Int32 id, Object inputArg, Object& outputArg)
   at Microsoft.VisualStudio.Html.Interop.CommandTargetToOleShim.Exec(Guid& guidCommandGroup, UInt32 commandID, UInt32 commandExecOpt, IntPtr variantIn, IntPtr variantOut)
   at Microsoft.VisualStudio.Editor.Implementation.CommandChainNode.InnerExec(Guid& pguidCmdGroup, UInt32 nCmdID, UInt32 nCmdexecopt, IntPtr pvaIn, IntPtr pvaOut)
   at Microsoft.VisualStudio.Editor.Implementation.SimpleTextViewWindow.Exec(Guid& pguidCmdGroup, UInt32 nCmdID, UInt32 nCmdexecopt, IntPtr pvaIn, IntPtr pvaOut)
   at Microsoft.VisualStudio.Editor.Implementation.CompoundTextViewWindow.Exec(Guid& pguidCmdGroup, UInt32 nCmdID, UInt32 nCmdexecopt, IntPtr pvaIn, IntPtr pvaOut)
   at Microsoft.VisualStudio.Platform.WindowManagement.DocumentObjectSite.Exec(Guid& pguidCmdGroup, UInt32 nCmdID, UInt32 nCmdexecopt, IntPtr pvaIn, IntPtr pvaOut)
   at Microsoft.VisualStudio.Platform.WindowManagement.WindowFrame.Exec(Guid& pguidCmdGroup, UInt32 nCmdID, UInt32 nCmdexecopt, IntPtr pvaIn, IntPtr pvaOut)
```

While trying to create a minimal reproduction (see `Index.cshtml` above), VS bugged me all the time with an messagebox, telling me `You cannot rename this element` (translated from german).
After restarting Visual Studio it was able to rename it without problems.

Unfortunately i couldn't reproduce it and also the renaming in the originally affected solution worked after restarting Visual Studio, so i'm opening this issue in the hope that someone will figure it out eventually.

EDIT: but i just run into it again so it seems to happen after some time.

### PR DESCRIPTION
**Pull request by dpoeschl**
Fixes #7660 
tag @dotnet/roslyn-ide

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioDocumentNavigationService.cs [hunk 1]
```diff
@@ -104,7 +108,10 @@ public bool TryNavigateToSpan(Workspace workspace, DocumentId documentId, TextSp
             }
 
             var text = document.GetTextAsync(CancellationToken.None).WaitAndGetResult(CancellationToken.None);
-            var textBuffer = text.Container.GetTextBuffer();
+            if (textSpan.Start < 0 || textSpan.End > text.Length)
+            {
+                return false;
```
Comment (DustinCampbell) [inline src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioDocumentNavigationService.cs]: I wonder if it'd be better to just make the span fit.

**Comment by jasonmalinowski** [comment]
Why are we getting bad spans in the first place? This looks like masking of a symptom, not fixing the root cause.

**Comment by dpoeschl** [comment]
@jasonmalinowski Indeed, we don't know. There were other instances of this ArgumentOutOfRangeException that came through different code paths where we just protected against the crash, I can dig it up if needed. I think one was done by @heejaechang. I can make this more clear and add a comment to that effect.

**Comment by jasonmalinowski** [comment]
Do we not have a repro of this? Do we know how widespread the crashes are?

**Comment by heejaechang** [comment]
did I mask some real bad issue? by the way, regardless how bad it is, wouldn't it be better if we save non-fatal watson and not crash VS?

**Comment by dpoeschl** [comment]
Let's first see if we catch the underlying problem with the cache validation stuff

**Pull request closed without merging**